### ISSUE TITLE
composer keeps using stale metadata

### ISSUE THREAD

**Issue by stefk** [composer/composer#5762]
My `composer.json`:
<details>
<summary>CLICK TO EXPAND</summary>
```json
{
    "name": "claroline/claroline",
    "description": "Claroline Connect",
    "license": "GPL-3.0+",
    "require": {
        "php": ">=5.5",
        "claroline/distribution": "dev-master"
    },
    "require-dev": {
        "fabpot/php-cs-fixer": "^1.10",
        "mikey179/vfsStream": "~1.5",
        "phpunit/phpunit": "~4.6",
        "symfony/phpunit-bridge": "*",
        "mockery/mockery": "^0.9.4",
        "phpmd/phpmd": "@stable",
        "stefk/md": "dev-master"
    },
    "scripts": {
        "fast-install" : [
            "php scripts/check.php",
            "php app/console doctrine:database:create --if-not-exists",
            "php vendor/bin/doctrine dbal:import claroline.sql -vvv",
            "php app/console assets:install web --symlink"
        ],
        "sync": [
            "composer fetch",
            "composer build",
            "php app/console claroline:update --env=prod",
            "composer disable-maintenance"
        ],
        "sync-dev": [
            "composer fetch-dev",
            "composer build",
            "php app/console claroline:update -vvv",
            "composer disable-maintenance"
        ],
        "fetch": [
            "composer install --prefer-dist --no-dev --optimize-autoloader",
            "npm install",
            "npm run bower"
        ],
        "fetch-dev": [
            "composer install --prefer-source",
            "npm install",
            "npm run bower"
        ],
        "build": [
            "Claroline\\CoreBundle\\Library\\Installation\\Refresher::deleteCache",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::buildBootstrap",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::installAssets",
            "npm run themes",
            "npm run webpack"
        ],
        "disable-maintenance": [
            "Claroline\\CoreBundle\\Library\\Maintenance\\MaintenanceHandler::disableMaintenance"
        ],
        "bundles": [
            "Claroline\\BundleRecorder\\ScriptHandler::buildBundleFile"
        ],
        "pre-install-cmd": [
            "php scripts/check.php",
            "php scripts/save-repo.php"
        ],
        "post-install-cmd": [
            "composer bundles"
        ],
        "pre-update-cmd": [
            "php scripts/check.php",
            "php scripts/save-repo.php",
            "Claroline\\CoreBundle\\Library\\Maintenance\\MaintenanceHandler::enableMaintenance"
        ],
        "post-update-cmd": [
            "composer bundles"
        ]
    },
    "minimum-stability": "dev",
    "prefer-stable": true,
    "config": {
        "github-protocols": ["https", "git", "ssh"],
        "process-timeout": 3600
    },
    "extra": {
        "symfony-app-dir": "app",
        "symfony-web-dir": "web",
        "symfony-assets-install": "symlink"
    }
}
```

</details>

Output of `composer diagnose`:
<details>
<summary>CLICK TO EXPAND</summary>
```
Checking platform settings: OK
Checking git settings: OK
Checking http connectivity to packagist: OK
Checking https connectivity to packagist: OK
Checking github.com oauth access: OK
Checking disk free space: OK
Checking pubkeys: 
Tags Public Key Fingerprint: 57815BA2 7E54DC31 7ECC7CC5 573090D0  87719BA6 8F3BB723 4E5D42D0 84A14642
Dev Public Key Fingerprint: 4AC45767 E5EC2265 2F0C1167 CBBB8A2B  0C708369 153E328C AD90147D AFE50952
OK
Checking composer version: OK
```
</details>

When I run any of these commands:
```
composer update
composer update claroline/distribution
```

the expected version of package `claroline/distribution` is installed but the metadata written in *composer.lock* and *vendor/composer/installed.json* are outdated, except for the commit references. 

In particular, the `autoload` section is stale, which in this case makes the whole package unusable since that section has been updated recently. Example: [autoload error](https://travis-ci.org/claroline/Claroline/builds/165198160#L2722) with an [installed version](https://travis-ci.org/claroline/Claroline/builds/165198160#L483) having a correct [autoload section](https://github.com/claroline/Distribution/blob/a15a298d1cc85d00f4fc85691b4b581cdfa5144e/composer.json#L198).

I thought it might be related to packagist but the data returned by the API is up-to-date and so is my local cache.

This is really a blocking issue, any help would be greatly appreciated.

**Comment by stefk** [comment]
It turns out that if the lock file is deleted before launching update commands, the new lock file and the local repository end up with up-to-date metadata, but that's obviously not a solution in the long run. I think that kind of partial updates should be considered as a bug. cc @alcohol

**Comment by stefk** [comment]
To help reproducing the issue, here's a lock file that will content outdated sections for `claroline/distribution` after an update: https://github.com/claroline/Claroline/blob/aef6e68485c709be2eff98180ef9e5899153a2de/composer.lock

**Comment by alcohol** [comment]
Sorry, but unable to reproduce:

```
$ wget -q https://raw.githubusercontent.com/claroline/Claroline/aef6e68485c709be2eff98180ef9e5899153a2de/composer.{json,lock}

$ ll
total 280K
-rw-r--r-- 1 rob users 2.8K Oct 18 08:24 composer.json
-rw-r--r-- 1 rob users 273K Oct 18 08:24 composer.lock

$ composer update --ignore-platform-reqs
> php scripts/check.php
Could not open input file: scripts/check.php
Script php scripts/check.php handling the pre-update-cmd event returned with error code 1
```

**Comment by alcohol** [comment]
...holy cow that is a lot of dependencies...

```
$ composer update --ignore-platform-reqs --no-scripts
Loading composer repositories with package information
Updating dependencies (including require-dev)
...
  - Installing claroline/front-end-bundle (dev-7.x-test 2baad17)
    Cloning 2baad174be4bcb35c87f8f5df07127de1d23fa17 from cache

  - Installing claroline/distribution (dev-master 21bbb9d)
    Cloning 21bbb9d4f28686d8c7f39ce8aa025992c9c582c0 from cache
...

$ composer show
...
claroline/distribution                   dev-master 21bbb9d   Claroline distribution
claroline/front-end-bundle               dev-7.x-test 2baad17 Front-end dependencies bundle for Claroline 2.0
...

$ grep -C6 'claroline/distribution' composer.lock
...
        {
            "name": "claroline/distribution",
            "version": "dev-master",
            "source": {
                "type": "git",
                "url": "https://github.com/claroline/Distribution.git",
                "reference": "21bbb9d4f28686d8c7f39ce8aa025992c9c582c0"
            },
...
```

Looks fine to me. I am closing this issue, until you can explain to us what you mean by "outdated sections" and on what exactly you base these claims.

**Comment by stefk** [comment]
Thanks for taking the time to try and reproduce the issue. This seems somehow related to the execution of scripts but I can't track down the exact problem right now. I'll come back with a simpler reproducible setup as soon as I can. As for the "outdated sections" mention, here's a simplified example of what I meant:

- A package x at version 1 has an autoload section containing entries for "a" and "b"
- Once I installed that package, the lock file references it correctly, including the autoload section with "a" and "b"
- The version 2 of the package adds an autoload entry "c"
- After updating with composer, the lock file references the version 2 correctly (I can see the git commit reference has been updated) BUT the autoload section in that file still contains only "a" and "b", not "c"

I have tons of build logs from different machines attesting of that behaviour. But indeed as soon as I deviate from the exact context of those builds, everything works as expected. Strange problem...
